cmds/init/init.go

```diff
@@ -169,6 +169,17 @@ func main() {
 		"/buildbin/rush",
 	}
 
+	// We need to reap the kids. This needs to start before the
+	// loop below since we might Run something that never returns.
+	go func() {
+		for {
+			var s syscall.WaitStatus
+			var r syscall.Rusage
+			if p, err := syscall.Wait4(-1, &s, 0, &r); err != nil {
```

Comment:
That should not happen. Init (pid 1) should only get Wait events for processes that have been reparented to it. So if, e.g., we Run a shell below, and that shell hangs around, the children of that shell will send exit status to the shell, not pid 1. That's part of the reason unix kernels maintain parent pid info -- so we know where to send exit status. When a process dies, its children typically get adopted (for exit purposes) by pid 1.

Hence pid 1 only reaps exit status of children for which it is the direct parent, or children whose parents have died. This wait will not preempt children who still have a real parent.

Reply:
Maybe I do not fully understand the mechanics of wait4, so here's an MVP:

Run the following with "sudo unshare -pf ./main" and it's output includes "exec.Command error 'wait: no child processes'" a non-deterministic number of times. It is a bit paradoxical because "exec.Command("true").Run()'' should always return 0. It seems like the exit status is being "stolen" from the os/exec. It looks like the same issue would occur when executing lines 205-207 which also check the exit status of os/exec.

```
package main

import (
        "log"
        "os"
        "os/exec"
        "syscall"
)

func main() {
        log.Printf("PID: %v\n", os.Getpid())

        // Kill children asynchronously.
        go func() {
                for {
                        var s syscall.WaitStatus
                        var r syscall.Rusage
                        if p, err := syscall.Wait4(-1, &s, 0, &r); err != nil {
                                log.Printf("%v: exited with %v, status %v", p, err, s)
                        }
                }
        }()

        // Run 10x to make the race condition more likely.
        for i := 0; i < 10; i++ {
                // You'd expect running this shell to always exit with 0.
                if err := exec.Command("true").Run(); err == nil {
                        log.Printf("exec.Command success\n")
                } else {
                        log.Printf("exec.Command error '%s'\n", err)
                }
        }
}
```

In fact, the issue brought up here isn't due to reparenting or PID 1,

The issue is if the shell below ever returns, the init process will be unable to reliably get the exit status.

Reply:
For reference, here's the output on my machine. Notice, that sometimes exec.Command is success and sometimes error:

```
2018/05/24 14:27:25 PID: 1
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 exec.Command error 'waitid: no child processes'
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 exec.Command success
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 exec.Command error 'wait: no child processes'
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 exec.Command error 'wait: no child processes'
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 -1: exited with no child processes, status 0
2018/05/24 14:27:25 exec.Command error 'wait: no child processes'
2018/05/24 14:27:25 -1: exited with no child processes, status 0
```

Reply:
I think something like the following should work. It correctly gets the exit status of the shell while reaping children:

```
package main

import (
        "log"
        "os"
        "os/exec"
        "syscall"
)

func main() {
        log.Printf("PID: %v\n", os.Getpid())

        cmd := exec.Command("/bin/sh")
        cmd.Stdin = os.Stdin
        cmd.Stdout = os.Stdout
        cmd.Stderr = os.Stderr
        if err := cmd.Start(); err != nil {
                log.Fatal("Error: ", err)
        }

        for {
                var s syscall.WaitStatus
                var r syscall.Rusage
                if p, err := syscall.Wait4(-1, &s, 0, &r); p == cmd.Process.Pid {
                        log.Printf("Shell exited, exit status %d", s.ExitStatus())
                        break
                } else if p != -1 {
                        log.Printf("Reaped PID %d, exit status %d", p, s.ExitStatus())
                } else {
                        log.Print("Error: ", err)
                        break
                }
        }
        if err := cmd.Process.Release(); err != nil {
                log.Fatal("Error: ", err)
        }
}
```

Example:

```
2018/05/25 00:34:44 PID: 1
# ( exit 42 & )
# 2018/05/25 00:34:50 Reaped PID 8, exit status 42

# exit 123
/bin/sh: 3: Cannot set tty process group (No such process)
2018/05/25 00:34:57 Shell exited, exit status 123
```

Reply:
ok, point taken, when you mentioned stealing I thought you were talking about the kernel, not the Go runtime. Nice to have you reviewing code again, that's a very good point :-)

OK, I'll rework this with your fix.

Reply:
darn, this won't quite do it. Once you've run all the commands, and exited the loop, there are probably still children you want to wait for before finally exiting.
I think if I add the reaper loop after I've run all the commands that might work, however.

Reply:
No problem, it is very subtle bug. :-)

I also just noticed busybox's init has a `waitfor` function to do what I suggested (reap children while waiting for a specific pid to exit): https://git.busybox.net/busybox/tree/init/init.c#n580

You can call a function like this for each single command. If any process gets reparented in-between commands, they will be reaped by the next command's call to `waitfor`. After the last command and when PID 1 exits, Linux automatically kills all the remaining processes in the pid namespace.